src/analysis/processing/qgsalgorithmrectanglesovalsdiamonds.cpp

```diff
@@ -0,0 +1,232 @@
+
+  mHeight = parameterAsDouble( parameters, QStringLiteral( "HEIGHT" ), context );
+  mDynamicHeight = QgsProcessingParameters::isDynamic( parameters, QStringLiteral( "HEIGHT" ) );
+  if ( mDynamicHeight )
+    mHeightProperty = parameters.value( QStringLiteral( "HEIGHT" ) ).value< QgsProperty >();
+
+  mRotation = parameterAsDouble( parameters, QStringLiteral( "ROTATION" ), context );
+  mDynamicRotation = QgsProcessingParameters::isDynamic( parameters, QStringLiteral( "ROTATION" ) );
+  if ( mDynamicRotation )
+    mRotationProperty = parameters.value( QStringLiteral( "ROTATION" ) ).value< QgsProperty >();
+
+  mSegments = parameterAsDouble( parameters, QStringLiteral( "SEGMENTS" ), context );
+
+  return true;
+}
+
+QgsFeatureList QgsRectanglesOvalsDiamondsAlgorithm::processFeature( const QgsFeature &feature, QgsProcessingContext &context, QgsProcessingFeedback * )
+{
+  QgsFeature outFeature = feature;
+  if ( outFeature.hasGeometry() )
+  {
+    QgsGeometry geometry = outFeature.geometry();
+    if ( geometry.isMultipart() )
+    {
+      throw QgsProcessingException( QObject::tr( "Multipart geometry. Please promote input layer to singleparts first." ) );
+    }
+
+    double width = mWidth;
+    if ( mDynamicWidth )
+      width = mWidthProperty.valueAsDouble( context.expressionContext(), width );
+
+    double height = mHeight;
+    if ( mDynamicHeight )
+      height = mHeightProperty.valueAsDouble( context.expressionContext(), height );
+
+    double rotation = mRotation;
+    if ( mDynamicRotation )
+      rotation = mRotationProperty.valueAsDouble( context.expressionContext(), rotation );
+
+    if ( width == 0 || height == 0 )
+    {
+      throw QgsProcessingException( QObject::tr( "Width and height should be greater than 0." ) );
+    }
+
+    double phi = rotation * M_PI / 180;
+    double xOffset = width / 2.0;
+    double yOffset = height / 2.0;
+    QgsPointXY point = geometry.asPoint();
+    double x = point.x();
+    double y = point.y();
+
+    QVector< double > ringX( 5 );
+    QVector< double > ringY( 5 );
+
+    switch ( mShape )
+    {
+      case 0:
+        // rectangle
+        ringX = { -xOffset, -xOffset, xOffset, xOffset, -xOffset };
+        ringY = { -yOffset, yOffset, yOffset, -yOffset, -yOffset };
+        break;
+      case 1:
+        // diamond
+        ringX = { 0.0, -xOffset, 0.0, xOffset, 0.0 };
+        ringY = { -yOffset, 0.0, yOffset, 0.0, -yOffset };
+        break;
+      case 2:
+        // oval
+        ringX.resize( mSegments + 1 );
+        ringY.resize( mSegments + 1 );
+        for ( int i = 0; i < mSegments; i ++ )
+        {
+          double t = ( 2 * M_PI ) / mSegments * i;
+          ringX[ i ] = xOffset * cos( t );
+          ringY[ i ] = yOffset * sin( t );
+        }
+        ringX[ mSegments ] = ringX.at( 0 );
+        ringY[ mSegments ] = ringY.at( 0 );
+        break;
+    }
+
+    for ( int i = 0; i < ringX.size(); ++i )
+    {
```

Comment:
Micro optimisation: if Phi is zero, we should skip this loop

Reply:
Done